micro_http/src/connection.rs

```diff
@@ -0,0 +1,714 @@
+            for cursor in 0..(end_cursor - line_start_index) {
+                self.buffer[cursor] = self.buffer[line_start_index + cursor];
+            }
+
+            // Clear the rest of the buffer.
+            for cursor in (end_cursor - line_start_index)..end_cursor {
+                self.buffer[cursor] = 0;
+            }
+        }
+
+        // Update `read_cursor`.
+        self.read_cursor = end_cursor - line_start_index;
+    }
+
+    /// Returns the first parsed request in the queue.
+    pub fn pop_parsed_request(&mut self) -> Option<Request> {
+        self.parsed_requests.pop_front()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::Version;
+    use super::*;
+    use std::os::unix::net::UnixStream;
+
+    #[test]
+    #[allow(clippy::cognitive_complexity)]
+    fn test_try_read_connection() {
+        // Test request with `Expect` header.
+        if let Ok((mut sender, receiver)) = UnixStream::pair() {
+            receiver.set_nonblocking(true).expect("Can't modify socket");
+            let mut conn = HttpConnection::new(receiver);
+            sender
+                .write_all(
+                    b"PATCH http://localhost/home HTTP/1.1\r\n \
+                                     Expect: 100-continue\r\n \
+                                     Content-Length: 26\r\n \
+                                     Transfer-Encoding: chunked\r\n\r\n",
+                )
+                .unwrap();
+            assert!(conn.try_read().is_ok());
+
+            sender.write_all(b"this is not\n\r\na json \nbody").unwrap();
+            conn.try_read().unwrap();
+            let request = conn.pop_parsed_request().unwrap();
+            let expected_request = Request::try_from(
+                b"PATCH http://localhost/home HTTP/1.1\r\n \
+                                     Expect: 100-continue\r\n \
+                                     Transfer-Encoding: chunked\r\n \
+                                     Content-Length: 26\r\n\r\nthis is not\n\r\na json \nbody",
+            )
+            .unwrap();
+            assert_eq!(request, expected_request);
+        }
+
+        // Long request headers.
+        if let Ok((mut sender, receiver)) = UnixStream::pair() {
+            receiver.set_nonblocking(true).expect("Can't modify socket");
+            let mut conn = HttpConnection::new(receiver);
+            sender
+                .write_all(
+                    b"PATCH http://localhost/home HTTP/1.1\r\n \
+                                     Expect: 100-continue\r\n \
+                                     Transfer-Encoding: chunked\r\n",
+                )
+                .unwrap();
+
+            for _ in 0..90 {
+                sender.write_all(b"Transfer-Encoding: chunked\r\n").unwrap();
+            }
+            sender
+                .write_all(b"Content-Length: 26\r\n\r\nthis is not\n\r\na json \nbody")
+                .unwrap();
+            assert!(conn.try_read().is_ok());
```

Comment:
Can you please explain why you assert twice here, and then you do a try_read().unwrap()?